2.segment_images/scripts/4.post-hoc_mask_refinement.py

```diff
@@ -0,0 +1,413 @@
+
+
+# In[ ]:
+
+
+if compartment == "nuclei":
+    mask_path = mask_dir / "nuclei_masks_reconstructed.tiff"
+    mask_output_path = mask_dir / "nuclei_masks_reconstructed_corrected.tiff"
+elif compartment == "cell":
+    mask_path = mask_dir / "cell_masks_watershed.tiff"
+    mask_output_path = mask_dir / "cell_masks_corrected.tiff"
+elif compartment == "organoid":
+    mask_path = mask_dir / "organoid_masks_reconstructed.tiff"
+    mask_output_path = mask_dir / "organoid_masks_reconstructed_corrected.tiff"
+
+else:
+    raise ValueError("Compartment must be either nuclei, cell or organoid")
+
+mask = tifffile.imread(mask_path)
+
+
+# ### Functions for refinement
+
+# In[4]:
+
+
+from typing import List, Tuple
+
+
+def calculate_bbox_area(bbox: Tuple[int, int, int, int]) -> int:
+    """
+    Calculate the area of a bounding box.
+
+    Parameters
+    ----------
+    bbox : Tuple[int, int, int, int]
+        The bounding box coordinates in the format (x_min, y_min, x_max, y_max).
+
+    Returns
+    -------
+    int
+        The area of the bounding box.
+    """
+    return (bbox[2] - bbox[0]) * (bbox[3] - bbox[1])
+
+
+def calculate_overlap(
+    bbox1: Tuple[int, int, int, int], bbox2: Tuple[int, int, int, int]
+) -> float:
+    # calculate the % overlap of the second bbox with the first bbox
+    if calculate_bbox_area(bbox1) == 0 or calculate_bbox_area(bbox2) == 0:
+        return 0.0
+    if calculate_bbox_area(bbox1) >= calculate_bbox_area(bbox2):
+        x_min = max(bbox1[0], bbox2[0])
+        y_min = max(bbox1[1], bbox2[1])
+        x_max = min(bbox1[2], bbox2[2])
+        y_max = min(bbox1[3], bbox2[3])
+        overlap_width = max(0, x_max - x_min)
+        overlap_height = max(0, y_max - y_min)
+        overlap_area = overlap_width * overlap_height
+        bbox1_area = calculate_bbox_area(bbox1)
+        bbox2_area = calculate_bbox_area(bbox2)
+        overlap_percentage = overlap_area / bbox2_area if bbox2_area > 0 else 0
+        return overlap_percentage
+    elif calculate_bbox_area(bbox1) < calculate_bbox_area(bbox2):
+        x_min = max(bbox1[0], bbox2[0])
+        y_min = max(bbox1[1], bbox2[1])
+        x_max = min(bbox1[2], bbox2[2])
+        y_max = min(bbox1[3], bbox2[3])
+        overlap_width = max(0, x_max - x_min)
+        overlap_height = max(0, y_max - y_min)
+        overlap_area = overlap_width * overlap_height
+        bbox1_area = calculate_bbox_area(bbox1)
+        bbox2_area = calculate_bbox_area(bbox2)
+        overlap_percentage = overlap_area / bbox1_area if bbox1_area > 0 else 0
+        return overlap_percentage
+    else:
+        print("Error: Bboxes are the same size")
+
+
+def merge_sets(list_of_sets: list) -> list:
+    for i, set1 in enumerate(list_of_sets):
+        for j, set2 in enumerate(list_of_sets):
+            if i != j and len(set1.intersection(set2)) > 0:
+                set1.update(set2)
+    return list_of_sets
+
+
+def check_for_all_same_labels(
```

Comment:
Consider making the function name here indicative to the boolean it provides - e.g. `includes_all_same_labels`. This might have the benefit of making the label use read more logically.